lib/Doctrine/ORM/Mapping/AssociationMapping.php

```diff
@@ -0,0 +1,240 @@
+    public static function fromMappingArray(array $mappingArray): static
+    {
+        unset($mappingArray['isOwningSide']);
+        $mapping = new static(
+            $mappingArray['fieldName'],
+            $mappingArray['sourceEntity'],
+            $mappingArray['targetEntity'],
+        );
+        foreach ($mappingArray as $key => $value) {
+            if (in_array($key, ['type', 'fieldName', 'sourceEntity', 'targetEntity'])) {
+                continue;
+            }
+
+            if ($key === 'joinTable') {
+                assert($mapping instanceof ManyToManyAssociationMapping);
+
+                if ($value === [] || $value === null) {
+                    continue;
+                }
+
+                assert($mapping instanceof ManyToManyOwningSideMapping);
+
+                $mapping->joinTable = JoinTableMapping::fromMappingArray($value);
+
+                continue;
+            }
+
+            if (property_exists($mapping, $key)) {
+                $mapping->$key = $value;
+            } else {
+                throw new Exception('Unknown property ' . $key . ' on class ' . static::class);
+            }
+        }
+
+        return $mapping;
+    }
+
+    /** @psalm-assert-if-true AssociationOwningSideMapping $this */
+    public function isOwningSide(): bool
+    {
+        return $this instanceof AssociationOwningSideMapping;
+    }
+
+    /** @psalm-assert-if-true ToOneAssociationMapping $this */
+    public function isToOne(): bool
+    {
+        return $this instanceof ToOneAssociationMapping;
+    }
+
+    /** @psalm-assert-if-true OneToOneOwningSideMapping|ManyToOneAssociationMapping $this */
+    public function isToOneOwningSide(): bool
+    {
+        return $this->isToOne() && $this->isOwningSide();
+    }
+
+    /** @psalm-assert-if-true ManyToManyOwningSideMapping $this */
+    public function isManyToManyOwningSide(): bool
+    {
+        return $this instanceof ManyToManyOwningSideMapping;
+    }
+
+    public function type(): int
+    {
+        return match (true) {
+            $this instanceof OneToOneAssociationMapping => ClassMetadata::ONE_TO_ONE,
+            $this instanceof OneToManyAssociationMapping => ClassMetadata::ONE_TO_MANY,
+            $this instanceof ManyToOneAssociationMapping => ClassMetadata::MANY_TO_ONE,
+            $this instanceof ManyToManyAssociationMapping => ClassMetadata::MANY_TO_MANY,
+            default => throw new Exception('Cannot determine type for ' . $this::class),
+        };
+    }
+
+    public function offsetGet($offset): mixed
+    {
+        if ($offset === 'isOwningSide') {
+            return $this->isOwningSide();
+        }
+
+        if ($offset === 'type') {
+            return $this->type();
+        }
+
+        return $this->$offset;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function offsetSet($offset, $value): void
+    {
+        if ($offset === 'joinTable') {
+            assert($this instanceof ManyToManyAssociationMapping);
```

Comment:
Why? I think we should avoid new exception for temporary code such as this one. It has no equivalent for properties.

Reply:
So the plan is to remove the `ArrayAccess` implementations before the release? In that case, I'm fine with the assert. Otherwise we need to assume that the array access is used downstream and this assert is basically input validation which is not what assert should be used for.

Reply:
No, the plan is to stop using array access internally before the release, to deprecate it in 3.1, and to remove it in 4.0.

I'll remove this assert, is was just a way for me to make sure I was doing this refactoring properly (checking my assumptions).